docs/source/matchers_tutorial.ipynb

```diff
@@ -0,0 +1,377 @@
+    "            if updated_node.value == \"False\":\n",
+    "                return updated_node.with_changes(value=\"True\")\n",
+    "        return updated_node\n"
+   ]
+  },
+  {
+   "cell_type": "raw",
+   "metadata": {
+    "raw_mimetype": "text/restructuredtext"
+   },
+   "source": [
+    "We can try it out on a source file to see that it works:\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "source = \"def some_func(*params: object) -> None:\\n    pass\\n\\nsome_func(True, False)\\nsome_func(1, 2, 3)\\nsome_func()\\n\"\n",
+    "module = cst.parse_module(source)\n",
+    "print(source)\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "new_module = module.visit(BoolInverter())\n",
+    "print(new_module.code)\n"
+   ]
+  },
+  {
+   "cell_type": "raw",
+   "metadata": {
+    "raw_mimetype": "text/restructuredtext"
+   },
+   "source": [
+    "While this works its not super elegant. We have to track where we are in the tree so we know when its safe to invert boolean literals which means we have to create a constructor and we have to duplicate matching logic. We could refactor that into a helper like the ``best_is_call_with_booleans`` above, but it only makes things so much better.\n",
+    "\n",
+    "So, lets try rewriting it with matcher decorators instead:\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class BetterBoolInverter(m.MatcherDecoratableTransformer):\n",
+    "    @m.call_if_inside(m.Call(args=(\n",
+    "        m.ZeroOrMore(m.Arg(m.Name(\"True\") | m.Name(\"False\"))),\n",
+    "    )))\n",
+    "    def leave_Name(self, original_node: cst.Name, updated_node: cst.Name) -> cst.Name:\n",
+    "        if updated_node.value == \"True\":\n",
+    "            return updated_node.with_changes(value=\"False\")\n",
+    "        if updated_node.value == \"False\":\n",
+    "            return updated_node.with_changes(value=\"True\")\n",
+    "        return updated_node\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "new_module = module.visit(BetterBoolInverter())\n",
+    "print(new_module.code)\n"
+   ]
+  },
+  {
+   "cell_type": "raw",
+   "metadata": {
+    "raw_mimetype": "text/restructuredtext"
+   },
+   "source": [
+    "Using matcher decorators we successfully removed all of the boilerplate around state tracking! The only thing that ``leave_Name`` needs to concern itself with is the actual business logic of the transform. However, it still needs to check to see if the value of the node should be inverted. This is because the ``Call.func`` is a ``Name`` in this case. Lets use another matcher decorator to make that problem go away:"
```

Comment:
Link Call.func and Name to its own documents.

Reply:
Good idea.